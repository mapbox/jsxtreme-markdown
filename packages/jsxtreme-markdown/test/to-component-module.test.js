'use strict';

const crypto = require('crypto');
const stripIndent = require('strip-indent');
const path = require('path');
const del = require('del');
const mkdirp = require('mkdirp');
const beautify = require('js-beautify');
const fs = require('fs');
const React = require('react');
const ReactDOMServer = require('react-dom/server');
const { promisify } = require('util');
const toComponentModule = require('../lib/to-component-module');

const prepText = text => stripIndent(text).trim();

// This directory needs to be within the project, so Babel will compile it and
// it has access to node_modules; but outside of Jest's roots, so the adding
// and deleting of files from it will not put Jest's watch mode in an endless
// loop. This is the reason we have Jests roots setting in package.json.
const tmpDir = path.join(__dirname, '../../../tmp');

const loadOutputModule = content => {
  const filename = path.join(
    tmpDir,
    crypto.randomBytes(16).toString('hex') + '.js'
  );

  return promisify(fs.writeFile)(filename, content).then(() => {
    let m = require(filename);
    m = m.default || m;
    return m;
  });
};

const renderComponent = (Component, props) => {
  return beautify.html(
    ReactDOMServer.renderToStaticMarkup(React.createElement(Component, props))
  );
};

describe('toComponentModule', () => {
  beforeAll(() => {
    return mkdirp(tmpDir);
  });

  afterAll(() => {
    return del(tmpDir);
  });

  afterEach(() => {
    return del(path.join(tmpDir, '*.*'));
  });

  test('default options', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # {{ frontMatter.title }}

      And a **special** number: {{ props.number }}.
    `);

    const code = toComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('default options produce valid module', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # {{ frontMatter.title }}

      And a **special** number: {{ props.number }}.
    `);

    const code = toComponentModule(text);
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output, {
        number: 77
      });
      expect(rendered).toMatchSnapshot();
    });
  });

  test('default options with prependJs in front matter', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      prependJs:
        - "const Timer = require('./timer')"
        - "const Watcher = require('./watcher').Watcher"
      ---

      # {{ frontMatter.title }}

      Some introductory text.

      {{ <Watcher /> }}

      This paragraph includes a {{ <Timer /> }}.
    `);

    const code = toComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('options.name', () => {
    const text = prepText(`
      # Title
    `);
    const options = {
      name: 'my-special-name'
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('options.template', () => {
    const text = prepText(`
      ---
      title: Foo
      list:
        - one
        - two
      ---

      # {{ frontMatter.title }}
    `);
    const options = {
      template: data => {
        return `${data.name}\n${JSON.stringify(data.frontMatter)}\n${data.jsx}`;
      }
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('toJsx options', () => {
    const text = prepText(`
      ---
      title: Foo
      list:
        - one
        - two
      ---

      # {# frontMatter.title #}

      {# <div className="foo"> #}
      The default {{ delimiter }} is ignored.
      {# </div> #}

      \`\`\`js
      var x = 3;
      \`\`\`
    `);
    const options = {
      delimiters: ['{#', '#}']
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('non-string primitives in front matter', () => {
    const text = prepText(`
      ---
      count: 7
      isHonest: false
      ---

      count: {{ frontMatter.count }}

      isHonest: {{ frontMatter.isHonest }}
    `);
    const code = toComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('options.wrapper', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      quantity: 834
      ---

      # {{ frontMatter.title }}

      Some introductory text. The quantity is {{ frontMatter.quantity }}

      Here is a number: {{ props.number }}
    `);
    const options = {
      wrapper: path.join(__dirname, './fixtures/wrapper.js')
    };
    const code = toComponentModule(text, options);
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output, {
        number: 77
      });
      expect(rendered).toMatchSnapshot();
    });
  });

  test('options.wrapper with ES2015 default export', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      quantity: 834
      ---

      # {{ frontMatter.title }}

      Some introductory text. The quantity is {{ frontMatter.quantity }}

      Here is a number: {{ props.number }}
    `);
    const options = {
      wrapper: path.join(__dirname, './fixtures/wrapper-es2015.js')
    };
    const code = toComponentModule(text, options);
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output, {
        number: 77
      });
      expect(rendered).toMatchSnapshot();
    });
  });

  test('options.prependJs', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      {{<Watcher />}}
    `);
    const options = {
      prependJs: ["import { Watcher } from './watcher'"]
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
  });

  test('documentation example, with wrapper front matter', () => {
    const text = prepText(`
      ---
      wrapper: '../wrapper'
      prependJs:
        - "import Timer from './timer'"
        - "import { Watcher } from './watcher'"
      title: Everything is ok
      quantity: 834
      ---

      # {{ frontMatter.title }}

      Some introductory text. The quantity is {{ frontMatter.quantity }}

      {{ <Watcher /> }}

      This paragraph includes a {{ <Timer /> }}.

      This component also accepts a "foo" prop: {{ props.foo }}
    `);
    const code = toComponentModule(text);
    expect(code).toMatchSnapshot();
  });

  test('options.precompile = true', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # {{ frontMatter.title }}

      Some introductory text.
    `);
    const options = {
      precompile: true
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output);
      expect(rendered).toMatchSnapshot();
    });
  });

  test('options.headings', () => {
    const text = prepText(`
      ---
      title: Everything is ok
      ---

      # First heading

      Some introductory text.

      ## Two
      ## Two
      ### Third heading, great
      # With {{ 3 }} interpolation
    `);
    const options = {
      headings: true
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output);
      expect(rendered).toMatchSnapshot();
    });
  });

  test('options.headings README example', () => {
    const text = prepText(`
      # One

      Text.

      ## Two

      Some more text.

      ### Third-level heading

      Yet more.

      ## Two

      A section with a duplicate title.
    `);
    const options = {
      headings: true
    };
    const code = toComponentModule(text, options);
    expect(code).toMatchSnapshot();
    return loadOutputModule(code).then(Output => {
      const rendered = renderComponent(Output);
      expect(rendered).toMatchSnapshot();
    });
  });
});
