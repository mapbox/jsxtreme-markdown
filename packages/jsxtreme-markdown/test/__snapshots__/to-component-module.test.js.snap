// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toComponentModule default options 1`] = `
"
    /*---
title: Everything is ok
---*/
    import React from 'react';
    
    const frontMatter = {
	title: 'Everything is ok'
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><h1>{frontMatter.title}</h1>
<p>And a <strong>special</strong> number: {props.number}.</p></div>;
      }
    }
  "
`;

exports[`toComponentModule default options produce valid module 1`] = `
"<div>
    <h1>Everything is ok</h1>
    <p>And a <strong>special</strong> number: 77.</p>
</div>"
`;

exports[`toComponentModule default options with prependJs in front matter 1`] = `
"
    /*---
title: Everything is ok
prependJs:
  - \\"const Timer = require('./timer')\\"
  - \\"const Watcher = require('./watcher').Watcher\\"
---*/
    import React from 'react';
    const Timer = require('./timer')
const Watcher = require('./watcher').Watcher

    const frontMatter = {
	title: 'Everything is ok'
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><h1>{frontMatter.title}</h1>
<p>Some introductory text.</p>
<Watcher />
<p>This paragraph includes a <Timer />.</p></div>;
      }
    }
  "
`;

exports[`toComponentModule documentation example, with wrapper front matter 1`] = `
"
    /*---
wrapper: '../wrapper'
prependJs:
  - \\"import Timer from './timer'\\"
  - \\"import { Watcher } from './watcher'\\"
title: Everything is ok
quantity: 834
---*/
    import React from 'react';
    import Timer from './timer'
import { Watcher } from './watcher'
import Wrapper from '../wrapper';

    const frontMatter = {
	title: 'Everything is ok',
	quantity: 834
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <Wrapper {...props} frontMatter={frontMatter}><div><h1>{frontMatter.title}</h1>
<p>Some introductory text. The quantity is {frontMatter.quantity}</p>
<Watcher />
<p>This paragraph includes a <Timer />.</p>
<p>This component also accepts a \\"foo\\" prop: {props.foo}</p></div></Wrapper>;
      }
    }
  "
`;

exports[`toComponentModule non-string primitives in front matter 1`] = `
"
    /*---
count: 7
isHonest: false
---*/
    import React from 'react';
    
    const frontMatter = {
	count: 7,
	isHonest: false
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><p>count: {frontMatter.count}</p>
<p>isHonest: {frontMatter.isHonest}</p></div>;
      }
    }
  "
`;

exports[`toComponentModule options.headings 1`] = `
"
    /*---
title: Everything is ok
---*/
    import React from 'react';
    
    const frontMatter = {
	title: 'Everything is ok',
	headings: [
		{
			text: 'First heading',
			slug: 'first-heading',
			level: 1
		},
		{
			text: 'Two',
			slug: 'two',
			level: 2
		},
		{
			text: 'Two',
			slug: 'two-1',
			level: 2
		},
		{
			text: 'Third heading, great',
			slug: 'third-heading-great',
			level: 3
		},
		{
			text: 'With i88 interpolation',
			slug: 'with-i88-interpolation',
			level: 1
		}
	]
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><h1 id=\\"first-heading\\">First heading</h1>
<p>Some introductory text.</p>
<h2 id=\\"two\\">Two</h2>
<h2 id=\\"two-1\\">Two</h2>
<h3 id=\\"third-heading-great\\">Third heading, great</h3>
<h1 id=\\"with-i88-interpolation\\">With {3} interpolation</h1></div>;
      }
    }
  "
`;

exports[`toComponentModule options.headings 2`] = `
"<div>
    <h1 id=\\"first-heading\\">First heading</h1>
    <p>Some introductory text.</p>
    <h2 id=\\"two\\">Two</h2>
    <h2 id=\\"two-1\\">Two</h2>
    <h3 id=\\"third-heading-great\\">Third heading, great</h3>
    <h1 id=\\"with-i88-interpolation\\">With 3 interpolation</h1>
</div>"
`;

exports[`toComponentModule options.headings README example 1`] = `
"
    
    import React from 'react';
    
    const frontMatter = {
	headings: [
		{
			text: 'One',
			slug: 'one',
			level: 1
		},
		{
			text: 'Two',
			slug: 'two',
			level: 2
		},
		{
			text: 'Third-level heading',
			slug: 'third-level-heading',
			level: 3
		},
		{
			text: 'Two',
			slug: 'two-1',
			level: 2
		}
	]
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><h1 id=\\"one\\">One</h1>
<p>Text.</p>
<h2 id=\\"two\\">Two</h2>
<p>Some more text.</p>
<h3 id=\\"third-level-heading\\">Third-level heading</h3>
<p>Yet more.</p>
<h2 id=\\"two-1\\">Two</h2>
<p>A section with a duplicate title.</p></div>;
      }
    }
  "
`;

exports[`toComponentModule options.headings README example 2`] = `
"<div>
    <h1 id=\\"one\\">One</h1>
    <p>Text.</p>
    <h2 id=\\"two\\">Two</h2>
    <p>Some more text.</p>
    <h3 id=\\"third-level-heading\\">Third-level heading</h3>
    <p>Yet more.</p>
    <h2 id=\\"two-1\\">Two</h2>
    <p>A section with a duplicate title.</p>
</div>"
`;

exports[`toComponentModule options.name 1`] = `
"
    
    import React from 'react';
    
    const frontMatter = {};

    export default class MySpecialName extends React.PureComponent {
      render() {
        const props = this.props;
        return <h1>Title</h1>;
      }
    }
  "
`;

exports[`toComponentModule options.precompile = true 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports[\\"default\\"] = void 0;

var _react = _interopRequireDefault(require(\\"react\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

function _typeof(obj) { \\"@babel/helpers - typeof\\"; if (typeof Symbol === \\"function\\" && typeof Symbol.iterator === \\"symbol\\") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === \\"function\\" && obj.constructor === Symbol && obj !== Symbol.prototype ? \\"symbol\\" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== \\"function\\" && superClass !== null) { throw new TypeError(\\"Super expression must either be null or a function\\"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === \\"object\\" || typeof call === \\"function\\")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError(\\"this hasn't been initialised - super() hasn't been called\\"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === \\"undefined\\" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === \\"function\\") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var frontMatter = {
  title: 'Everything is ok'
};

var MarkdownReact = /*#__PURE__*/function (_React$PureComponent) {
  _inherits(MarkdownReact, _React$PureComponent);

  var _super = _createSuper(MarkdownReact);

  function MarkdownReact() {
    _classCallCheck(this, MarkdownReact);

    return _super.apply(this, arguments);
  }

  _createClass(MarkdownReact, [{
    key: \\"render\\",
    value: function render() {
      var props = this.props;
      return /*#__PURE__*/_react[\\"default\\"].createElement(\\"div\\", null, /*#__PURE__*/_react[\\"default\\"].createElement(\\"h1\\", null, frontMatter.title), /*#__PURE__*/_react[\\"default\\"].createElement(\\"p\\", null, \\"Some introductory text.\\"));
    }
  }]);

  return MarkdownReact;
}(_react[\\"default\\"].PureComponent);

exports[\\"default\\"] = MarkdownReact;"
`;

exports[`toComponentModule options.precompile = true 2`] = `
"<div>
    <h1>Everything is ok</h1>
    <p>Some introductory text.</p>
</div>"
`;

exports[`toComponentModule options.prependJs 1`] = `
"
    /*---
title: Everything is ok
---*/
    import React from 'react';
    import { Watcher } from './watcher'

    const frontMatter = {
	title: 'Everything is ok'
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <Watcher />;
      }
    }
  "
`;

exports[`toComponentModule options.template 1`] = `
"MarkdownReact
{\\"title\\":\\"Foo\\",\\"list\\":[\\"one\\",\\"two\\"]}
<h1>{frontMatter.title}</h1>"
`;

exports[`toComponentModule options.wrapper 1`] = `
"<div>
    <div>This is the wrapper. Here are its props:</div>
    <div>77</div>
    <div>Everything is ok</div>
    <div>834</div>
    <div>
        <h1>Everything is ok</h1>
        <p>Some introductory text. The quantity is 834</p>
        <p>Here is a number: 77</p>
    </div>
</div>"
`;

exports[`toComponentModule options.wrapper with ES2015 default export 1`] = `
"<div>
    <div>This is the wrapper. Here are its props:</div>
    <div>77</div>
    <div>Everything is ok</div>
    <div>834</div>
    <div>
        <h1>Everything is ok</h1>
        <p>Some introductory text. The quantity is 834</p>
        <p>Here is a number: 77</p>
    </div>
</div>"
`;

exports[`toComponentModule toJsx options 1`] = `
"
    /*---
title: Foo
list:
  - one
  - two
---*/
    import React from 'react';
    
    const frontMatter = {
	title: 'Foo',
	list: [
		'one',
		'two'
	]
};

    export default class MarkdownReact extends React.PureComponent {
      render() {
        const props = this.props;
        return <div><h1>{frontMatter.title}</h1>
<div className=\\"foo\\">
<p>The default {\\"{\\"}{\\"{\\"} delimiter {\\"}\\"}{\\"}\\"} is ignored.</p>
</div>
<pre><code className=\\"language-js\\">var x = 3;{\\"\\\\n\\"}</code></pre></div>;
      }
    }
  "
`;
